File: solutions/python/spiral-matrix/2/spiral_matrix.py
"""Spiral matrix generator."""

from itertools import cycle

DIRECTIONS = ((0, 1), (1, 0), (0, -1), (-1, 0))


def spiral_matrix(size):
    """Return a square matrix filled in clockwise spiral order."""
    matrix = [[None] * size for _ in range(size)]
    directions = cycle(DIRECTIONS)
    row = 0
    column = 0
    row_delta, column_delta = next(directions)

    for value in range(1, size * size + 1):
        matrix[row][column] = value

        if should_turn(matrix, size, (row, column), (row_delta, column_delta)):
            row_delta, column_delta = next(directions)

        row += row_delta
        column += column_delta

    return matrix


def should_turn(matrix, size, position, direction):
    """Return whether the next cursor movement needs to turn."""
    row, column = position
    row_delta, column_delta = direction
    next_row = row + row_delta
    next_column = column + column_delta

    return (
        not 0 <= next_row < size
        or not 0 <= next_column < size
        or matrix[next_row][next_column] is not None
    )

File: solutions/scala/spiral-matrix/2/src/main/scala/SpiralMatrix.scala
object SpiralMatrix:
  def spiralMatrix(size: Int): List[List[Int]] =
    if size == 0 then List.empty
    else
      val matrix = Array.fill(size, size)(0)
      var top = 0
      var bottom = size - 1
      var left = 0
      var right = size - 1
      var number = 1

      while top <= bottom && left <= right do
        for column <- left to right do
          matrix(top)(column) = number
          number += 1
        top += 1

        for row <- top to bottom do
          matrix(row)(right) = number
          number += 1
        right -= 1

        if top <= bottom then
          for column <- right to left by -1 do
            matrix(bottom)(column) = number
            number += 1
          bottom -= 1

        if left <= right then
          for row <- bottom to top by -1 do
            matrix(row)(left) = number
            number += 1
          left += 1

      matrix.map(_.toList).toList
